simpleflow/canvas.py

```diff
@@ -0,0 +1,125 @@
+from . import futures
+from .activity import Activity
+from .task import ActivityTask
+
+
+class FuncGroup(object):
+    """
+    Class calling a function returning an ActivityTask, a group or a chain
+    activities : Group, Chain...
+    """
+    def __init__(self, func, *args, **kwargs):
+        self.func = func
+        self.args = list(args)
+        self.kwargs = kwargs
+
+    def submit(self, executor):
+        inst = self.func(*self.args, **self.kwargs)
+        if not isinstance(inst, (ActivityTask, Group)):
+            raise TypeError('FuncGroup submission should return a Group or an ActivityTask, Got {} instead'.format(type(inst)))
+        return inst.submit(executor)
+
+
+class Group(object):
+    """
+    List of activities running in parallel.
+    """
+
+    def __init__(self,
+                 *activities,
+                 **options):
+        self.activities = list(activities)
+
+    def append(self, *args, **kwargs):
+        if isinstance(args[0], (ActivityTask, Group)):
+            self.activities.append(args[0])
+        elif isinstance(args[0], Activity):
+            self.activities.append(ActivityTask(*args, **kwargs))
+        else:
+            raise ValueError('{} should be an ActivityTask or an Activity'.format(args[0]))
+
+    def submit(self, executor):
+        return GroupFuture(self.activities, executor)
+
+
+class GroupFuture(futures.Future):
+
+    def __init__(self, activities, executor):
+        self.activities = activities
+        self.futures = []
+        self.executor = executor
+        self._state = futures.PENDING
+        for a in self.activities:
+            self.futures.append(self._submit_activity(a))
+        self.sync_state()
+        self.sync_result()
+
+    def _submit_activity(self, act):
+        if isinstance(act, ActivityTask):
+            return self.executor.submit(act.activity, *act.args, **act.kwargs)
+        elif isinstance(act, (Group, FuncGroup)):
+            return act.submit(self.executor)
+        raise TypeError('Bad type for `act` ({}). Waiting for `ActivityTask`, `Group` or `FuncGroup` instead'.format(type(act)))
+
+    def sync_state(self):
+        if all(a.finished for a in self.futures):
```

Comment:
I added it. If at least one activity is cancelled, consider the whole group cancelled.